otherlibs/dune-rpc/private/procedures.ml

```diff
@@ -118,12 +118,48 @@ module Poll = struct
   let name t = t.name
 
   module Progress = struct
+    module V1 = struct
+      open Progress
+
+      let sexp =
+        let open Conv in
+        let waiting = constr "waiting" unit (fun () -> Waiting) in
+        let failed = constr "failed" unit (fun () -> Failed) in
+        let in_progress =
+          let complete = field "complete" (required int) in
+          let remaining = field "remaining" (required int) in
+          constr "in_progress"
+            (record (both complete remaining))
+            (fun (complete, remaining) ->
+              In_progress { complete; remaining; failed = 0 })
+        in
+        let interrupted = constr "interrupted" unit (fun () -> Interrupted) in
+        let success = constr "success" unit (fun () -> Success) in
+        let constrs =
+          List.map ~f:econstr [ waiting; failed; interrupted; success ]
+          @ [ econstr in_progress ]
+        in
+        let serialize = function
+          | Waiting -> case () waiting
+          | In_progress { complete; remaining; failed = _ } ->
+            case (complete, remaining) in_progress
+          | Failed -> case () failed
+          | Interrupted -> case () interrupted
+          | Success -> case () success
+        in
+        sum constrs serialize
+    end
+
     let name = "progress"
 
     let v1 =
+      Decl.Request.make_current_gen ~req:Id.sexp ~resp:(Conv.option V1.sexp)
```

Comment:
If I understand the versioning mechanism correctly, this should be a call to `Decl.Request.make_gen`, which takes additional parameters to upgrade/downgrade to/from the current generation's types. I assume that `make_current_gen` _would_ work just fine, but the disadvantage is that the serialization and upgrade logic is now intertwined. With `make_gen`, we can leave the old sexp function exactly the way it was, and have a small, isolated bit of upgrade and downgrade logic.

To be concrete, I'm suggesting that the old version of `Progress.t` live in the above `V1` module, and the extra parameters to `make_gen` should define how to go back and forth between `Progress.t` and `V1.t`.

Reply:
I'm not sure I understand what this achieves. If we introduce the old type in V1, then we have two choices for `v1` either it is a `Progress.t` in which case now we will need the old sexp converter anyway, in addition to new conversion functions. Otherwise, we can make `v1` have `V1.t option`, but then the list of versions is no longer homogeneous.

Reply:
I guess it does separate the conversion and serialization code however.

Reply:
Oops, just checked my email and saw these messages; github hid them after the commit was pushed. 

Yeah, the separation was all I was after, and the reason I care about that is that each time a new version is introduced, we have to go back and update all prior versions as well. With the separation, we can compose v1_to_v2 and v2_to_v3 to get v1_to_v3 for free; I think it's also less error-prone if we have the rule that we never touch the serialization code of past versions.